drivers/video/goldfish_fb.c

```diff
@@ -0,0 +1,436 @@
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <debug.h>
+#include <nuttx/config.h>
+#include <nuttx/video/fb.h>
+#include <nuttx/irq.h>
+#include <nuttx/kmalloc.h>
+#include <nuttx/mm/circbuf.h>
+
+/****************************************************************************
+ * Pre-processor definitions
+ ****************************************************************************/
+
+#ifndef putreg32
+#define putreg32(v, x) (*(volatile uint32_t *)(x) = (v))
+#endif
+
+#ifndef getreg32
+#define getreg32(x) (*(uint32_t *)(x))
+#endif
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+enum
+{
+  GOLDFISH_FB_GET_WIDTH = 0x00,
+  GOLDFISH_FB_GET_HEIGHT = 0x04,
+  GOLDFISH_FB_INT_STATUS = 0x08,
+  GOLDFISH_FB_INT_ENABLE = 0x0c,
+  GOLDFISH_FB_SET_BASE = 0x10,
+  GOLDFISH_FB_SET_ROTATION = 0x14,
+  GOLDFISH_FB_SET_BLANK = 0x18,
+  GOLDFISH_FB_GET_PHYS_WIDTH = 0x1c,
+  GOLDFISH_FB_GET_PHYS_HEIGHT = 0x20,
+  GOLDFISH_FB_GET_FORMAT = 0x24,
+  GOLDFISH_FB_INT_VSYNC = 1U << 0,
+  GOLDFISH_FB_INT_UPDATE_DONE = 1U << 1,
+  GOLDFISH_FB_FORMAT_BRGA_8888 = 1,
+  GOLDFISH_FB_FORMAT_RGBX_8888 = 2,
+  GOLDFISH_FB_FORMAT_RGB_888 = 3,
+  GOLDFISH_FB_FORMAT_RGB_565 = 4,
+  GOLDFISH_FB_FORMAT_BGRA_8888 = 5,
+  GOLDFISH_FB_FORMAT_RGBA_5551 = 6,
+  GOLDFISH_FB_FORMAT_RGBA_4444 = 8
+};
+
+struct goldfish_fb_format_s
+{
+  uint8_t fmt;
+  uint8_t bpp;
+};
+
+struct goldfish_fb_s
+{
+  struct fb_vtable_s vtable;
+  struct fb_planeinfo_s planeinfo;
+  struct fb_videoinfo_s videoinfo;
+  FAR void *base;
+  int irq;
+  struct circbuf_s vsync;   /* Vsync event queued */
+#ifdef CONFIG_GOLDFISH_FB_VIDEO_MODE
+  bool busy; /* Only used in the video mode */
+  uintptr_t cur_buf;
+#endif
+};
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static FAR struct goldfish_fb_s *g_goldfish_fb;
+
+static const struct goldfish_fb_format_s g_goldfish_format_map[] =
```

Comment:
remove static and move g_goldfish_format_map table inside function up_fbinitialize() to save the code size

Reply:
Done.